### PR TITLE
[JENKINS-31155] Shared libraries

### PR DESCRIPTION
**Pull request by jglick**
[JENKINS-31155](https://issues.jenkins-ci.org/browse/JENKINS-31155)

- [x] global configuration option
- [x] per-folder configuration option
- [X] `@Library` annotation
- [x] basic SCM checkout & load
- [x] test https://github.com/jenkinsci/git-plugin/pull/433
- [x] global variable support (incl. test of [JENKINS-32400](https://issues.jenkins-ci.org/browse/JENKINS-32400) scenario)
- [x] _Global Variables Reference_ support
- [x] Replay support
- [x] form validation for library definitions
- [x] resource loading
- [x] restart tests
- [x] `checkpoint` tests
- [x] `@Grab` integration test (presupposes #9)
- [x] basic interactive tests
- [x] dynamic GitHub `LibraryResolver`: https://github.com/jenkinsci/github-organization-folder-plugin/pull/22
- [x] interactive tests using an organization folder: https://github.com/cloudbeers/PR-demo/pull/15
- [x] ~~`library` step exploration~~ (seems like a dead end)
- [X] inline help (needs https://github.com/jenkinsci/scm-api-plugin/pull/13)
- [x] [documentation](https://github.com/jglick/workflow-cps-global-lib-plugin/blob/shared-libs-JENKINS-31155/README.md)
- [X] https://github.com/jenkinsci/subversion-plugin/pull/168
- [X] update Pipeline-as-code demo to use it: https://github.com/jenkinsci/github-branch-source-plugin/pull/72
- [x] refactor API to use `LibraryRetriever`
- [x] make sure every library requested is used
- [x] fix deletion of last entry in `GlobalLibraries`

@reviewbybees

### CODE REVIEW

**Comment by jglick** [comment]
There is a ` TypeError: Cannot read property "value" from undefined` in `GlobalLibrariesTest.configRoundtrip` but it seems to go away in Jenkins 2 (or at least between 1.651.3 and 2.7.2) so I am not inclined to spend time on it. In the future I would like to fix the JS libraries to provide more meaningful errors.

**Comment by rsandell** [comment]
:bee:

File: README.md [hunk 1]
```diff
@@ -216,6 +402,32 @@ void parallelize(int count) {
 }
 ```
 
-Since the _Overall/RunScripts_ permission is required to push any changes to the global library repository,
-library code is allowed to call any Java methods without sandbox checks.
-Therefore it can safely encapsulate calls to third-party libraries (as well as to Jenkins model APIs, etc.).
+Libraries are cached by default in `~/.groovy/grapes/` on the Jenkins master.
```
Comment (rsandell) [inline README.md]: How will this work when you are running Jenkins as a "system account" in linux, i.e. a user that doesn't have a home directory?

Comment (jglick) [inline README.md]: Well `System.getProperty("user.home")` is going to return some directory, I do not offhand know what. Depends on the installation package.

File: README.md [hunk 2]
```diff
@@ -30,13 +35,67 @@ The basename of each `*.groovy` file should be a Groovy (~ Java) identifier, con
...
+and then use versions such as `trunk` or `branches/dev` or `tags/1.0`.
+
+### Global external libraries
+
+There are several places you can define external libraries, according to your needs.
+
+Under _Manage Jenkins » Configure System » Global Pipeline Libraries_ you may configure as many libraries as you need.
+These will be accessible to any Pipeline job in the system.
+
+These libraries are trusted: they can run any methods in Java, Groovy, Jenkins internal APIs, Jenkins plugins, or third-party libraries.
```
Comment (oleg-nenashev) [inline README.md]: :ant: IMHO they should be trusted if and only if the library definition contains commitId or other such value, which cannot be changed easily. Otherwise it implies a security risk in the case somebody gets an access to the repo wit the lib (e.g. version label can be moved to another commit).

File: README.md [hunk 3]
```diff
@@ -30,13 +35,67 @@ The basename of each `*.groovy` file should be a Groovy (~ Java) identifier, con
...
+This allows you to define libraries which encapsulate individually unsafe APIs in a higher-level wrapper safe for use from any job.
+Beware that **anyone able to push commits to this SCM repository could obtain unlimited access to Jenkins**.
```
Comment (oleg-nenashev) [inline README.md]: Works very bad for Enterprises, where you unlikely can do anything with it due to security policies and repo settings. 
/me was impacted by it many times

File: README.md (hunk 2 above)
Comment (jglick) [inline README.md]: Well it is just up to the Jenkins administrator to implement proper security controls on that repository. Using nondeterministic (e.g., branch) revisions on global libraries is an important use case.

The administrator already has the option of defining the library with a nonoverridable version pointing to a tag or (better) commit hash.

File: README.md (hunk 3 above)
Comment (jglick) [inline README.md]: No one is forcing the administrator to use, and trust, global libraries. They can simply decline to do that, and allow untrusted libraries to be defined at the folder level.

**Comment by jglick** [comment]
Sigh, snapshot problems…

File: README.md [hunk 4]
```diff
@@ -1,25 +1,30 @@
...
 
 When you have multiple Pipeline jobs, you often want to share some parts of the Pipeline
 scripts between them to keep Pipeline scripts [DRY](http://en.wikipedia.org/wiki/Don't_repeat_yourself).
 A very common use case is that you have many projects that are built in the similar way.
 
-This plugin adds that functionality by creating a "shared library script" Git repository inside Jenkins.
-Every Pipeline script in your Jenkins see these shared library scripts in their classpath.
+This plugin adds that functionality by allowing you to create “shared library script” SCM repositories.
+It can be used in two modes:
+a legacy mode in which there is a single Git repository hosted by Jenkins itself, to which you may push changes;
```
Comment (KostyaSha) [inline README.md]: Do  you mean your closed source plugin?

Comment (jglick) [inline README.md]: No. What closed source plugin?

Comment (KostyaSha) [inline README.md]: git repo for validated merges AFAIR. Or what example could be for such use case?

Comment (jglick) [inline README.md]: The CloudBees Validated Merge plugin? They both depend on `git-server` but that is the only relation I can see.

The note is just that this plugin can either be used with an internal repository, as before; or, now, and preferably, with an external repository.

Comment (KostyaSha) [inline README.md]: So internal would require placing some files somewhere in `JENKINS_HOME` because core doesn't provide some extension for having folders for plugin needs?

Comment (jglick) [inline README.md]: Yes, the internal mode since 1.0 was using a Jenkins-hosted Git repository.

**Comment by abayer** [comment]
:bee:

**Comment by jglick** [comment]
@abayer may have found a buglet in `LibraryAdder` (as yet unconfirmed): if a given library was requested, but no `LibraryResolver` actually provided a definition for it, an error should be reported rather than allowing execution to continue (and presumably fail later).

Not quite trivial since a resolver may return implicit libraries; so the Javadoc for `LibraryResolver` should clarify that if it wishes to “claim” a given library request, it must return a `LibraryConfiguration` with a matching `name` (which currently implementations in fact do). Probably `LibraryAdder` should also then avoid even asking subsequent resolvers about that library.

**Comment by jglick** [comment]
Thought of a possible alternate design for `LibraryConfiguration`: rather than take an `SCMSource`, include a field for some new extension point like `LibrarySource`, one implementation of which would take an `SCMSource` (implementing the new `retrieve` overload, reverting the patch to `SimpleSCMSource`), another of which would take an `SCM`. The extension point would define a method to populate the build’s `libs/$name/` directory, extracting this code from `LibraryAdder`.

Such a change would make the configuration UI for legacy SCMs a little cleaner, and make it simpler to add support for (say) retrieving libraries from an artifact repository in the future. Possible to do this compatibly retroactively but a bit more work since we would need a `readResolve` in `LibraryConfiguration`.

**Comment by abayer** [comment]
fwiw, a test for this is at https://github.com/abayer/pipeline-model-definition-plugin/blob/library-check/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/BasicModelDefTest.java#L51-L59 - the test should fail, since I just copy-pasted it, but the loaded function doesn't seem to be invoked at all, and the build in fact succeeds (the test fails due to things not showing up in the logs that it expects).

**Comment by HRMPW** [comment]
When trying to add a new shared library on Jenkins master (system config) I get an NPE when specifying "master" or "Master" branch to load implicitly with no credentials. Works fine when credentials are set.

{code}
Cannot validate default version.
java.lang.NullPointerException
	at com.cloudbees.plugins.credentials.matchers.IdMatcher.<init>(IdMatcher.java:56)
	at com.cloudbees.plugins.credentials.CredentialsMatchers.withId(CredentialsMatchers.java:132)
	at jenkins.plugins.git.AbstractGitSCMSource.getCredentials(AbstractGitSCMSource.java:423)
	at jenkins.plugins.git.AbstractGitSCMSource.doRetrieve(AbstractGitSCMSource.java:180)
	at jenkins.plugins.git.AbstractGitSCMSource.retrieve(AbstractGitSCMSource.java:357)
	at jenkins.scm.api.SCMSource.fetch(SCMSource.java:278)
	at org.jenkinsci.plugins.workflow.libs.LibraryConfiguration$DescriptorImpl.doCheckDefaultVersion(LibraryConfiguration.java:195)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.kohsuke.stapler.Function$InstanceFunction.invoke(Function.java:324)
	at org.kohsuke.stapler.Function.bindAndInvoke(Function.java:167)
	at org.kohsuke.stapler.Function.bindAndInvokeAndServeResponse(Function.java:100)
	at org.kohsuke.stapler.MetaClass$1.doDispatch(MetaClass.java:124)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:746)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:876)
	at org.kohsuke.stapler.MetaClass$5.doDispatch(MetaClass.java:233)
	at org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)
	at org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:746)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:876)
	at org.kohsuke.stapler.Stapler.invoke(Stapler.java:649)
	at org.kohsuke.stapler.Stapler.service(Stapler.java:238)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:812)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1669)
	at hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:135)
	at hudson.util.PluginServletFilter.doFilter(PluginServletFilter.java:126)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at hudson.security.csrf.CrumbFilter.doFilter(CrumbFilter.java:86)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:84)
	at hudson.security.UnwrapSecurityExceptionFilter.doFilter(UnwrapSecurityExceptionFilter.java:51)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:117)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.providers.anonymous.AnonymousProcessingFilter.doFilter(AnonymousProcessingFilter.java:125)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.rememberme.RememberMeProcessingFilter.doFilter(RememberMeProcessingFilter.java:142)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.ui.AbstractProcessingFilter.doFilter(AbstractProcessingFilter.java:271)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at jenkins.security.BasicHeaderProcessor.doFilter(BasicHeaderProcessor.java:93)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at org.acegisecurity.context.HttpSessionContextIntegrationFilter.doFilter(HttpSessionContextIntegrationFilter.java:249)
	at hudson.security.HttpSessionContextIntegrationFilter2.doFilter(HttpSessionContextIntegrationFilter2.java:67)
	at hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:87)
	at hudson.security.ChainedServletFilter.doFilter(ChainedServletFilter.java:76)
	at hudson.security.HudsonFilter.doFilter(HudsonFilter.java:171)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at org.kohsuke.stapler.compression.CompressionFilter.doFilter(CompressionFilter.java:49)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at hudson.util.CharacterEncodingFilter.doFilter(CharacterEncodingFilter.java:82)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at org.kohsuke.stapler.DiagnosticThreadNameFilter.doFilter(DiagnosticThreadNameFilter.java:30)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:585)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:553)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:223)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:499)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
	at winstone.BoundedExecutorService$1.run(BoundedExecutorService.java:77)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
{code}

**Comment by kohsuke** [comment]
:bee:

**Comment by kohsuke** [comment]
:bee:

**Comment by abayer** [comment]
Currently a :bug: on this - [this test](https://github.com/abayer/pipeline-model-definition-plugin/blob/ec4262022792ad41d8bf6a64a5917e2d469b8919/src/test/java/org/jenkinsci/plugins/pipeline/modeldefinition/BasicModelDefTest.java#L52) still doesn't actually execute anything and has no errors in the execution either, which is...bad. That's running against latest SNAPSHOTs of everything...

**Comment by jglick** [comment]
> this test

Using, I guess, [this branch](https://github.com/jenkinsci/pipeline-model-definition-plugin/compare/master...abayer:library-check). Will take a look.

**Comment by jglick** [comment]
> I get an NPE

https://github.com/jenkinsci/git-plugin/pull/433/commits/9cd577a27ff1319a39fd7b165ba8963c78632caa

**Comment by jglick** [comment]
@abayer seems to be a problem with named parameters together with a block parameter: passing one named parameter works; two does not. And using non-named-parameter syntax fails. Cannot reproduce outside `pipeline-model-definition`, so I suspect that plugin is just doing something too weird to Groovy. Working on producing a minimal test case for you, which passes only when `org.jenkinsci.plugins.pipeline.modeldefinition.parser.GroovyShellDecoratorImpl` is commented out, leading me to suspect that `ModelASTTreeStep` or similar is buggy. I do not see any reason to believe the problem is with this PR.

**Comment by abayer** [comment]
@jglick That is weird but plausible. I swear we have usage of two named parameters and a block in the tests, but I could be mistaken. Anyway, bug revoked, have a :bee:

**Comment by rsandell** [comment]
:bee:

File: src/main/java/org/jenkinsci/plugins/workflow/libs/SCMSourceRetriever.java [hunk 5]
```diff
@@ -75,13 +75,13 @@ public SCMSource getScm() {
...
-        doRetrieve(scm.build(revision.getHead(), revision), revision.isDeterministic(), target, run, listener);
+        doRetrieve(scm.build(revision.getHead(), revision), target, run, listener);
     }
 
-    static void doRetrieve(@Nonnull SCM scm, boolean deterministic, FilePath target, Run<?, ?> run, TaskListener listener) throws Exception {
+    static void doRetrieve(@Nonnull SCM scm, FilePath target, Run<?, ?> run, TaskListener listener) throws Exception {
         // Adapted from CpsScmFlowDefinition:
         SCMStep delegate = new GenericSCMStep(scm);
-        delegate.setPoll(!deterministic); // TODO is this desirable?
+        delegate.setPoll(false); // TODO we have no API for determining if a given SCMHead is branch-like or tag-like; would we want to turn on polling if the former?
```
Comment (rsandell) [inline src/main/java/org/jenkinsci/plugins/workflow/libs/SCMSourceRetriever.java]: IMHO The library is not the one being built, it is just a tool to achieve the build. So for me it is ok to never poll the library.
If I want to test the library I'll setup a separate job for that :)

**Comment by rsandell** [comment]
:bee:

**Comment by abayer** [comment]
re-:bee:

File: src/main/java/org/jenkinsci/plugins/workflow/libs/SCMSourceRetriever.java (hunk 5 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/libs/SCMSourceRetriever.java]: Yeah you could argue it either way. With a `checkout` step the user chooses, but the default is to poll. The argument to poll (when the SCM refers to something like a branch) is that if the library does change, and you were to do another build, it might behave differently, which is something you probably want to know earlier rather than later. On the other hand, you might be making a very minor change to a library with hundreds of usages.
